File: sdk-python/document_trust_payload/canonicalize.py
import json
from typing import Any


def _serialize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if not (value == value and value not in (float("inf"), float("-inf"))):
            raise TypeError("Payload contains a non-finite number")
        return json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(_serialize(item) for item in value) + "]"
    if isinstance(value, dict):
        parts = []
        for key in sorted(value.keys()):
            item = value[key]
            parts.append(f"{json.dumps(str(key), ensure_ascii=False)}:{_serialize(item)}")
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"Unsupported payload value type: {type(value).__name__}")


def canonicalize(value: Any) -> str:
    return _serialize(value)

File: sdk-js/src/canonicalize.js
function normalizeNumber(value) {
  if (!Number.isFinite(value)) {
    throw new TypeError("Payload contains a non-finite number");
  }
  return JSON.stringify(value);
}

function serialize(value) {
  if (value === null) {
    return "null";
  }

  switch (typeof value) {
    case "string":
      return JSON.stringify(value);
    case "number":
      return normalizeNumber(value);
    case "boolean":
      return value ? "true" : "false";
    case "object":
      if (Array.isArray(value)) {
        return `[${value.map((item) => serialize(item)).join(",")}]`;
      }

      {
        const keys = Object.keys(value).sort();
        const pairs = [];
        for (const key of keys) {
          const item = value[key];
          if (item === undefined) {
            continue;
          }
          pairs.push(`${JSON.stringify(key)}:${serialize(item)}`);
        }
        return `{${pairs.join(",")}}`;
      }
    default:
      throw new TypeError(`Unsupported payload value type: ${typeof value}`);
  }
}

export function canonicalize(value) {
  return serialize(value);
}
